File: src/main/java/com/l98293/phone/PhoneValidator.java
package com.l98293.phone;

import com.google.i18n.phonenumbers.NumberParseException;
import com.google.i18n.phonenumbers.PhoneNumberUtil;
import com.google.i18n.phonenumbers.Phonenumber.PhoneNumber;
import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

public class PhoneValidator implements ConstraintValidator<Phone, String> {

	private Region region;

	@Override
	public void initialize(Phone constraintAnnotation) {
		this.region = constraintAnnotation.region();
	}

	@Override
	public boolean isValid(String value, ConstraintValidatorContext context) {
		if (value == null || value.trim().isEmpty()) {
			return true;
		}

		PhoneNumberUtil phoneNumberUtil = PhoneNumberUtil.getInstance();
		try {

			PhoneNumber phonenumber = phoneNumberUtil.parse(value, region.getCode());
			return phoneNumberUtil.isValidNumber(phonenumber);
		} catch (NumberParseException e) {
			throw new RuntimeException(e);
		}
	}
}

File: src/main/java/com/l98293/phone/PhoneValidator.kt
package com.l98293.phone

import com.google.i18n.phonenumbers.PhoneNumberUtil
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class PhoneValidator: ConstraintValidator<Phone, String> {

    private lateinit var region: Region
    private lateinit var format: Format

    override fun initialize(constraintAnnotation: Phone) {

        region = constraintAnnotation.region
        format = constraintAnnotation.format
    }

    override fun isValid(
        value: String?,
        context: ConstraintValidatorContext?
    ) = value.isNullOrEmpty() || runCatching {

        val phoneNumberUtil = PhoneNumberUtil.getInstance()
        val phoneNumber = phoneNumberUtil.parseAndKeepRawInput(
            value,
            region.code
        )

        phoneNumberUtil.isValidNumberForRegion(
            phoneNumber,
            region.code
        ) && format.matches(phoneNumber.countryCodeSource)
    }.getOrDefault(false)
}